### PR TITLE
Addition of information on nuclei

### PR DESCRIPTION
**Pull request by eduardo-rodrigues**
Fixes https://github.com/scikit-hep/particle/issues/210.

While adding information on nuclei, which adds *a lot*, I encountered a serious slow down of `Particle.load_table`. Turns out that comes from a bug in its logic, see the fix `if append and value in cls._table:`. Making the check only run when relevant makes the loading > 10X faster :-). Now adding nucleus info won't any longer be an issue.

I will be adding the actual info shortly, with minor updates here and there.

FYI @henryiii and @HDembinski.

### CODE REVIEW

**Comment by HDembinski** [comment]
I am very happy to see this coming. I looked into the implementation and was surprised that there are only edits of one file. Is the DB already part of the repo?

**Comment by eduardo-rodrigues** [comment]
Nope. I need to make some modifications to `Particle` and update the nuclei CSV file. I'm also checking various bits ... but I basically have a working version with me. We really are almost there ...

**Comment by eduardo-rodrigues** [comment]
@HDembinski, I've just committed the file. I will need to add info on how it was created …

Would you be happy with this kind of output:
```
In [2]: print(Particle.from_pdgid(1000130270).describe())
Name: Al27           ID: 1000130270   Latex: $^{27}\mathrm{Al}$
Mass  = 25133.1437306 MeV
Width = ?
Q (charge)        = 13      J (total angular) = None     P (space parity) = ?
C (charge parity) = ?       I (isospin)       = None     G (G-parity)     = ?
    SpinType: SpinType.NonDefined
    Antiparticle name: Al27~ (antiparticle status: Barred)
```
?

**Comment by eduardo-rodrigues** [comment]
For info here is how the CSV file of nucleus information got created:

- Use list of nuclei (and their isotopes) from the `periodictable` package.
- Use its masses and convert them to MeV scaling everything to account for a small difference in the mass of the neutron as provided by `periodictable` compared to what we have, as in the PDG data.
- Copy all information for the neutron and proton but with their corresponging "nucleus PDG ID":
  neutron: 2112 <-> 1000000010
  proton: 2212 <-> 1000010010
  which means that neutrons and protons appear twice, for completeness.

**Comment by eduardo-rodrigues** [comment]
@henryiii, @HDembinski, I think this is now ready for discussion and review, modulo tests and minor things that will come later and separately.

**Comment by eduardo-rodrigues** [comment]
Oh, @HDembinski - by the time nuclei are released in the package I hope you will consider to give us a star of appreciation :D.

**Comment by eduardo-rodrigues** [comment]
As "background information": these updates to the PDG ID functions may at first seem strange, or an overkill, but they make things consistent given the special case of neutrons and protons in terms of PDG IDs. Indeed, atop the usual PDG IDs 2112 and 2212, also the "nucleus PDG IDs" 1000000010 and 1000010010 exist (see for example the `periodictable` package), and should be equivalent. This last update ensures this. (Admitedly it would have been easier if these special cases did not exist, but ...)

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -7,7 +7,6 @@ repos:
 - repo: https://github.com/pre-commit/pre-commit-hooks
   rev: v2.3.0
   hooks:
-  - id: check-added-large-files
```
Comment (henryiii) [inline .pre-commit-config.yaml]: Why are we dropping this? You can [add a larger max size]
(https://github.com/pre-commit/pre-commit-hooks#check-added-large-files).

File: particle/particle/particle.py [hunk 2]
```diff
@@ -378,8 +385,12 @@ def load_table(cls, filename=None, append=False):
             cls._table_names = []
 
         if filename is None:
-            filename = data.open_text(data, "particle2019.csv")
-            cls._table_names.append("particle2019.csv")
+            import fileinput
```
Comment (henryiii) [inline particle/particle/particle.py]: This should be imported at the top - importing inside a function should only be done for special cases, a standard library module is just fine at the top.

File: particle/particle/particle.py [hunk 3]
```diff
@@ -388,39 +399,46 @@ def load_table(cls, filename=None, append=False):
         else:
             cls._table_names.append("{0!r} {1}".format(filename, len(cls._table_names)))
 
-        with filename as f:
+        # The following 2 files are necessary for Python 2, otherwise simply
+        # with filename as f:
+        # works just fine in Python 3 !
+        from contextlib import closing
+
+        with closing(filename) as f:
```
Comment (henryiii) [inline particle/particle/particle.py]: I assume closing on something that is already a context manager is fine?

File: .pre-commit-config.yaml (hunk 1 above)
Comment (eduardo-rodrigues) [reply .pre-commit-config.yaml]: > Let's get the updated build system in before we actually push 0.10.0, too.

Sure, no problem. The fact that I updated the CHANGELOG does not mean that I would push a release out. I remember you saying you have things in the pipeline. I just noticed the CHANGELOG was rather out-of-date.

Comment (eduardo-rodrigues) [reply .pre-commit-config.yaml]: The file with all nuclei info is actually just over 500 KB, the limit. That line was not allowing me to commit :S. But it's most likely safe to re-add the line after this PR … (I had a quick look to see how to increase the limit to 1MB but did not immediately find a solution, hence went pragmatically as I just wanted to get the job done.)

File: particle/particle/particle.py (hunk 2 above)
Comment (eduardo-rodrigues) [reply particle/particle/particle.py]: OK, will get to this and the rest tomorrow. Let me call it a day - it's getting rather late.

Thanks for starting to take a look, BTW.

File: particle/particle/particle.py (hunk 3 above)
Comment (eduardo-rodrigues) [reply particle/particle/particle.py]: I found this workaround online and it seems it is precisely to solve thie kind of issue I got with Python 2 here.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (henryiii) [reply .pre-commit-config.yaml]: `args: ['--maxkb=600']` (placed just below the id)

Comment (eduardo-rodrigues) [reply .pre-commit-config.yaml]: Done, thanks.

File: particle/particle/particle.py (hunk 2 above)
Comment (eduardo-rodrigues) [reply particle/particle/particle.py]: Done putting the imports at the start of the file.

**Comment by eduardo-rodrigues** [comment]
Both happy?

**Comment by HDembinski** [comment]
Right, I gave you a star just now. :)

This looks all good. What happens when you do the old `Particle.find("p")`, do you get an error, because it is ambiguous now or does it return the 2212 proton? I think it should return the 2212 for backward compatibility.

**Comment by HDembinski** [comment]
```
In [2]: print(Particle.from_pdgid(1000130270).describe())
Name: Al27           ID: 1000130270   Latex: $^{27}\mathrm{Al}$
Mass  = 25133.1437306 MeV
Width = ?
Q (charge)        = 13      J (total angular) = None     P (space parity) = ?
C (charge parity) = ?       I (isospin)       = None     G (G-parity)     = ?
    SpinType: SpinType.NonDefined
    Antiparticle name: Al27~ (antiparticle status: Barred)
```
This looks good to me. Is "?" a placeholder?

**Comment by eduardo-rodrigues** [comment]
:-).

Yeah, so now it gives you
```
In [4]: Particle.findall(name="p")
Out[4]:
[<Particle: name="p", pdgid=2212, mass=938.272081 ± 0.000006 MeV>,
 <Particle: name="p", pdgid=1000010010, mass=938.272081 ± 0.000006 MeV>]
```
and `find` will tell you you have too many particles, since 2. This can be discussed but clearly one has 2 possibilities: either just have the proton (similar arguments for the neutron) just as 2212, or else have it identical but with PDG ID 1000010010, which is a valid ID used in various places for the H nucleu=proton, e.g. in `periodictable`. Personally I prefer to be more comprehensive, even if `Particle.findall(name="p")` now provides 2 replies.
I realise some documentation on nuclei needs to be included very soon ...

**Comment by eduardo-rodrigues** [comment]
> In [2]: print(Particle.from_pdgid(1000130270).describe())
> Name: Al27           ID: 1000130270   Latex: $^{27}\mathrm{Al}$
> Mass  = 25133.1437306 MeV
> Width = ?
> Q (charge)        = 13      J (total angular) = None     P (space parity) = ?
> C (charge parity) = ?       I (isospin)       = None     G (G-parity)     = ?
>     SpinType: SpinType.NonDefined
>     Antiparticle name: Al27~ (antiparticle status: Barred)
> 
> This looks good to me. Is "?" a placeholder?

It is difficult to have consistency everywhere - super annoying - given that we have all kinds of missing or non-relevant quantum numbers depending on what we're talking about (the actual particles, the specific internal quantum numbers). In general the "?" means the information is not known/available.
In some cases you get a `None` to be consistent with that is returned by the ID functions, which mimic the best way possible what the C++ HepPID package was providing (for completeness let me say that `Particle` has a much more complete and less buggy version of the functions at this stage ;-)).

**Comment by HDembinski** [comment]
Why not always return `None` when information is missing? This seems very Pythonic.

**Comment by eduardo-rodrigues** [comment]
A concrete example to what I mean: think of G-parity: not relevant for most particles. One could go fancy and make a distinction between not available and non-relevant, but I suspect it will be quite difficult to deal with all use-cases … Certainly something to think about in the near future, as I agree some more consistency will be less confusing. Thanks for the ping on this.

**Comment by eduardo-rodrigues** [comment]
> Why not always return None when information is missing? This seems very Pythonic.

So same as not relevant. @henryiii, any strong feelings? We can certainly try and make some changes, but then in a different PR ...

**Comment by henryiii** [comment]
Without looking into it very closely, I'm in agreement with information missing == None.

**Comment by eduardo-rodrigues** [comment]
Thanks both!

**Pull request merged**